contracts/GnosisSafePersonalEdition.sol

```diff
@@ -90,29 +157,93 @@ contract GnosisSafePersonalEdition is MasterCopy, GnosisSafe, SignatureValidator
     {
-        // There cannot be an owner with address 0.
-        address lastOwner = address(0);
-        address currentOwner;
-        uint256 i;
-        // Validate threshold is reached.
-        for (i = 0; i < threshold; i++) {
-            currentOwner = recoverKey(txHash, signatures, i);
-            require(owners[currentOwner] != 0, "Signature not provided by owner");
-            require(currentOwner > lastOwner, "Signatures are not ordered by owner address");
-            lastOwner = currentOwner;
+        signedMessage[getMessageHash(_data)] = 1;
+    }
+
+    /**
+    * @dev Should return whether the signature provided is valid for the provided data
+    * @param _data Arbitrary length data signed on the behalf of address(this)
+    * @param _signature Signature byte array associated with _data
+    * @return a bool upon valid or invalid signature with corresponding _data
+    */ 
+    function isValidSignature(bytes _data, bytes _signature)
+        public
+        view 
+        returns (bool isValid)
+    {
+        bytes32 messageHash = getMessageHash(_data);
+        if (_signature.length == 0) {
+            isValid = signedMessage[messageHash] != 0;
+        } else {
+            isValid = checkSignatures(messageHash, _data, _signature);
         }
     }
 
+    /// @dev Returns hash of a message that can be signed by owners.
+    /// @param message Message that should be hashed
+    /// @return Message hash.
+    function getMessageHash(
+        bytes message
+    )
+        public
+        view
+        returns (bytes32)
+    {
+        bytes32 safeMessageHash = keccak256(
+            abi.encode(SAFE_MSG_TYPEHASH, keccak256(message))
+        );
+        return keccak256(
+            abi.encodePacked(byte(0x19), byte(1), domainSeperator, safeMessageHash)
+        );
+    }
+
+    /// @dev Returns the bytes that are hashed to be signed by owners.
+    /// @param to Destination address.
+    /// @param value Ether value.
+    /// @param data Data payload.
+    /// @param operation Operation type.
+    /// @param safeTxGas Fas that should be used for the safe transaction.
+    /// @param dataGas Gas costs for data used to trigger the safe transaction.
+    /// @param gasPrice Maximum gas price that should be used for this transaction.
+    /// @param gasToken Token address (or 0 if ETH) that is used for the payment.
+    /// @param _nonce Transaction nonce.
+    /// @return Transaction hash bytes.
+    function encodeTransactionData(
+        address to, 
+        uint256 value, 
+        bytes data, 
+        Enum.Operation operation, 
+        uint256 safeTxGas, 
+        uint256 dataGas, 
+        uint256 gasPrice, 
+        address gasToken,
+        uint256 _nonce
+    )
+        public
+        view
+        returns (bytes)
+    {
+        bytes32 safeTxHash = keccak256(
+            abi.encode(SAFE_TX_TYPEHASH, to, value, keccak256(data), operation, safeTxGas, dataGas, gasPrice, gasToken, _nonce)
```

Comment:
Hashing data `keccak256(data)` is done to be compatible with EIP 1077?

Reply:
It is to make it compatible with EIP-712